### DISCUSSION TITLE
[Feature Request] Align how _inputNode is checked in Input and Textarea

### DISCUSSION THREAD

**Discussion by amielucha** [ing-bank/lion#579] [General] [answered]
While I was reviewing the differences between the implementation of property changing in [input](https://github.com/ing-bank/lion/blob/54f5853c1f2a63c291be61710245b0668ab4854f/packages/input/src/LionInput.js#L75) and [textarea](https://github.com/ing-bank/lion/blob/60284d79bc538632661e2f00ef47aeb191988a91/packages/textarea/src/LionTextarea.js#L78) components, I noticed there's a slightly different approach to how `this._inputNode` value is checked. In Textarea there's an additional variable `native`.

```javascript
const native = this._inputNode;
if (native) {
  native.readOnly = this.readOnly;
}
```

Is the check necessary? In which scenario native would be `undefined`? If it's needed, shouldn't it be also applied to `updated()` in Input?

```javascript
updated(changedProps) {
  super.updated(changedProps);
  if (changedProps.has('type')) {
    this._inputNode.type = this.type;
  }
  if (changedProps.has('placeholder')) {
    this._inputNode.placeholder = this.placeholder;
  }
}
```

**Comment by daKmoR** [comment]
the difference is that 

```js
const native = this._inputNode;
if (native) {
  native.readOnly = this.readOnly;
}
```
get's executed `sync` in  https://github.com/ing-bank/lion/blob/54f5853c1f2a63c291be61710245b0668ab4854f/packages/input/src/LionInput.js#L60

so this can get executed before the dom node is available (maybe it gets rendered?)

---

In comparison `updated` always gets executed *async* AFTER a render - therefore, in this case, we can be sure that the element is available.



does my explanation makes sense? 😅

**Comment by amielucha** [comment]
Thanks for the clarification. This raises additional question though.

If there is no need for the check in `updated`, could this

https://github.com/ing-bank/lion/blob/60284d79bc538632661e2f00ef47aeb191988a91/packages/textarea/src/LionTextarea.js#L68-L73

knowing that it's inside `updated` be simplified as follows?

```js
if (changedProperties.has('rows')) {
  this._inputNode.rows = this.rows;
}
```

If it's the case, I could follow up with another PR after the [addition of placeholders in Textarea](https://github.com/ing-bank/lion/pull/578).

**Comment by jorenbroekema** [comment]
Personally I wouldn't mind if we can get rid of redundant guards, so feel free to send a PR, if the tests pass I would be okay with merging, going by daKmoR's logic it should be fine. If it fails, it might be an indication something else is wrong.

**Comment by jorenbroekema** [comment]
@daKmoR  I looked into this more and my proposal would be the following:
- Have a protected delegation method in LionInput:
```js
_delegatePropToInputNode(propName) {
  this._inputNode[propName] = this[propName];
}
``` 
- Call it in connectedCallback() for synchronously delegating on initial (because slotmixin appends slots in connectedCallback, so then we do it immediately when the node is available)
- Call it in updated() if changedProperties has propName, for imperatively updating a prop and making sure it's synced down to the input node as well.

Then we make sure it's consistent everywhere, for rows, autocomplete, placeholder, type, readonly etc. etc.

Thoughts?

**Comment by jorenbroekema** [comment]
Closing as enhancement with votes needed. If someone wants to push for this, feel free! Or vote if you think it's important by 👍 'ing.

**Comment by jorenbroekema** [comment, marked as answer]
We have strict types now in Lion, and inputNode is now always strictly typed as not being undefined, so that kinda solves it and removes the need for the guard 👍
